docs/guides/contexts_expressions.md

```diff
@@ -0,0 +1,390 @@
+```
+
+Some example context key names (currently) include:
+
+| Name     | Description |
+| -------- | -------- |
+| `layers_selection_count`  | Number of layers currently selected  | 
+| `all_layers_linked`  | True when all selected layers are linked  | 
+| `active_layer_is_rgb`  | True when the active layer is RGB  | 
+| `active_layer_type`  | Lowercase name of active layer type, or None if no layer is active.  | 
+| `only_images_selected`  | True when there is at least one selected layer and all selected layers are images  | 
+| `active_layer_ndim` | Number of dimensions in the active layer, or `None` if nothing is active |
+...  many more
+
+
+### `ContextKey` objects
+
+To track the special **name** strings that can be used in expressions, napari
+has the `ContextKey` class.  Instances of `ContextKey` subclass from
+[`ast.Name`](https://docs.python.org/3/library/ast.html#ast.Name) and represent
+a variable name in an expression. Additionally, they have a `description` that
+can be used in documentation, a `default_value`, and a `getter` function that
+can be called to retrieve the current value (the parameters passed to the
+`getter` will depend on the context key... see [Updating
+Contexts](#updating-contexts) below for more).
+
+In order to keep related keys together, `ContextKey` instances will usually be
+declared as class attributes on a `ContextNamespace` class:
+
+```python
+# all of the getters here receive an instance of viewer.layers.selection
+
+class LayerListContextKeys(ContextNamespace):
+    layers_selection_count = ContextKey(
+        default_value=0,
+        description="Number of layers currently selected",
+        getter=lambda sel: len(sel),
+    )
+    active_layer_is_rgb = ContextKey(
+        default_value=False,
+        description="True when the active layer is RGB",
+        getter=lambda sel: getattr(sel.active, "rgb", False),
+    )
+    active_layer_ndim = ContextKey(
+        None,
+        "Number of dimensions in the active layer, or `None` if nothing is active",
+        lambda s: s.active and getattr(s.active.data, "ndim", None),
+    )
+```
+
+You can think of a `ContextNamespace` kind of like an `Enum` class.  And just
+like an `Enum`, you can see all of its members:
+
+```python
+In [13]: LayerListContextKeys.__members__
+Out[13]:
+mappingproxy({
+    'layers_selection_count': ContextKey(id='layers_selection_count'),
+    'active_layer_is_rgb': ContextKey(id='active_layer_is_rgb'),
+    'active_layer_ndim': ContextKey(id='active_layer_ndim')
+})
+```
+
+A nice aspect of `ContextKeys` is that they can be used in expressions.  But
+unlike a simple string, they can also provide type hinting, linting
+capabilities, and IDE autocompletion.
+
+```python
+In [14]: expr = LayerListContextKeys.active_layer_ndim >= 3 
+
+In [15]: expr.eval({'active_layer_ndim': 2})
+Out[15]: False
+```
+
+A record of all registered context keys can be retrieved with the class method
+`ContextKey.info()`
```

Comment:
I think this can be ambiguous as wether it's `info()` on an instance, or `info` class method on the supper class.